### ISSUE TITLE
raw ADO.NET and stored procs

### ISSUE THREAD

**Issue by Chicagoan2016** [MarimerLLC/cslaforum#361]
Hi All,
we are using raw ADO.NET and stored procs ( use of stored proc is department standard), however our data access code is extremely error prone, one simple mistake in typing parameter name and we get error.

Has anyone here faced similar situation?
Any suggestions will be appreciated.

kind regards

**Comment by tfreitasleal** [comment]
Hi,
Typing mistake are bound to happen whenever you don't use strong names - names the compiler can't check, like names in strings.
To prevent/reduce/miniize/eliminate this issue, you can use a code generator.
To make sure everything is correct, you can build tests: unit tests and full integration tests. The later should test database interaction and mean a lot of work, because your test procedures must initialize the database to known state. Nevertheless this is the only to test methods using ADO.NET and stored procedures. If you use databinding, you should also have UI tests to make sure binding strings are correct

**Comment by jonnybee** [comment]
Hi, 

I would use some codegen to generate wrappers over stored procedure with strong named parameters. This eliminates handcoding and also gives you easier insight, f.ex into which code call a given SP.

**Comment by GillesBer** [comment]
Hi all,
@Chicagoan2016 : we were facing the same troubles, and are now generating all the ADO.NET code to perform insert, updates, delete as .NET wrappers around each stored proc. 
We did not use standard code generator, but we implement a small executable (~700 LOC, using Microsoft.SqlServer.Management.Smo), and this is really a time saver (we are generating 11 k LOC for one of our apps).
In addition to avoid typos, we were able to implement additional checks (ex. that our strings were not longer than the varchar(n) DB type / to manage nulls / ..  ) and add additional auditing in debug mode. 

We did not manage the Select (ie. still use untyped sqlReader / dataTable). 
At that time, I found it too difficult to manage multi result sets. 
But if someone has some experience on managing result sets, feel free to share!

**Comment by ajj7060** [comment]
Code gen, using something like Entity Framework (which can call stored procedures to do its work).

**Comment by GillesBer** [comment]
@ajj7060  Thanks, it was not the case in older version of EF.
It seems to be possible now, https://msdn.microsoft.com/en-us/library/jj691402(v=vs.113).aspx
I will reconsider it.

**Comment by Chicagoan2016** [comment]
thank you all,
I appreciate your help.
@GillesBer what code generator do you use? Is there a way we could possibly purchase it from your company?
@ajj7060  I will look at EF.

@tfreitasleal if I am not mistaken you developed cslagenfork, do we have a tutorial about cslagenfork? 
we are using encapsulated Invoke.

kind regards

**Comment by GillesBer** [comment]
@Chicagoan2016 : This is 'hand made' (and probably not optimum), but I do think that - if you did it manually/yourself:
- this is quite easy for you to generate the code a way that it is easy to consume.
- you will have extra value by adding your custom rules.

I shared my code as an example on: https://github.com/GillesBer/CodeSample/blob/master/TemplateHelper.cs
It is ~700LOC written in C#, and uses the connection string as a parameter.
The code generated is in VB.NET; it is flushed to the Console.Out

ex
`
CREATE  Proc [dbo].[IU_CommentForDiagram] (
	@CommentId  numeric(10,0),
	@Comment  varchar(500)
                                     )
As
  Begin
...
  END`

will generate 

```
`Function Exec_IU_CommentForDiagram (ByVal vCommentId As Decimal, ByVal vComment As String) As Long
            Dim paramCommentId As New SqlParameter("@CommentId", SqlDbType.Int)
            paramCommentId.Value = vCommentId
            Dim paramComment As New SqlParameter("@Comment", SqlDbType.VarChar, 500)
            paramComment.Value = IIf(vComment Is Nothing, DBNull.Value, vComment.CheckLength(500,"IU_CommentForDiagram.Comment"))

            Dim lngRtrn As Long = ExecuteStoredProc("IU_CommentForDiagram", paramCommentId, paramComment )
                Return lngRtrn
            End Function`
```

You will have to provide your own implementation of ExecuteStoredProc (but you should already have this somewhere in your code) and a few extension methods such as String.CheckLength (there are for auding purpose).

**Comment by Chicagoan2016** [comment]
Thank You

**Comment by Chicagoan2016** [comment]
I was looking at an older project that we developed in 2005 using Data Access application block and it has methods like ExecuteReader and ExecuteNonQuery which takes parameters to stored proc as an object array. The most recent version of Enterprise library has same (similar) methods. If we use these methods from Enterprise library, does CSLA's connectionmanager work with Enterprise library? just to make sure we don't open a distributed transaction unnecessarily. 

kind regards

**Comment by GillesBer** [comment]
Sincerely, the implementation is so trivial (a few hours), that you/we benefit from moving out of the Data Access application block. Here is an example of our implementation:
```
      Private Function ExecuteStoredProc(CommandText As String, 
                                  commandParameters() As SqlParameter) As Long
                Dim lngAffected As Long
                Using scope As ICTransactionScope = NewTransactionScope()
                    Dim cmd As DbCommand = scope.NewCommand
                    With cmd
                        .CommandTimeout = ApplicationConfiguration.Current.CommandTimeout
                        .CommandText = CommandText
                        .CommandType = CommandType.StoredProcedure
                        .Parameters.AddRange(commandParameters)
                        Using (scope.AppendQuery(CommandText, commandParameters))
                            lngAffected = .ExecuteNonQuery()
                        End Using
                        .Parameters.Clear()
                    End With

                    scope.Complete()
                    Return lngAffected
                End Using

        End Function````

**Comment by jonnybee** [comment]
Just out of curiosity - why not have a Using on the DbCommand object (and then a Dispose) and do you really need Parameters.Clear()? 

Or is the command object owned by the ICTransactionScope and disposed in that context?

**Comment by GillesBer** [comment]
Good point, you are correct: this is not used in this context / I will comment it.
I exclude some of my code which performs audit and attempt to perform retry, the 'clear()' was there to ensure the code was reentrant.
